File: app/CausalityCSharp/CausalityLibrary/Type/Scene.cs
using System;

namespace CausalityLibrary.Type
{
    [Serializable]
    public sealed class Scene
    {
        public SerialNumber SerialNumber
        { get; }

        public string Description
        { get; }

        public SerialNumber Caption
        { get; }

        public bool HasCause
        { get; }

        public SerialNumber Cause
        { get; }

        public bool Displayed
        { get; set; }

        public Scene(SerialNumber serialNumber,
            string description, SerialNumber caption,
            bool displayed = false)
        {
            SerialNumber = serialNumber;
            Description = description;
            Caption = caption;
            Displayed = displayed;
        }
        public Scene(SerialNumber serialNumber,
            string description, SerialNumber caption,
            SerialNumber cause, bool displayed = false)
        {
            SerialNumber = serialNumber;
            Description = description;
            Caption = caption;
            Cause = cause;
            HasCause = true;
            Displayed = displayed;
        }

        // Don't instantiate with default constructor!
        private Scene() { }
    }
}

File: app/CausalityJava/src/com/inha/dsp/causality/type/Scene.java
package com.inha.dsp.causality.type;

import java.io.Serializable;

public class Scene implements Serializable {
    private SerialNumber serialNumber;
    private String description;
    private SerialNumber caption;
    private SerialNumber cause;
    private boolean hascause;

    public SerialNumber getSerialNumber() {
        return serialNumber;
    }
    public String getDescription() {
        return description;
    }
    public SerialNumber getCaption() {
        return caption;
    }
    public boolean hasCause() {
        return hascause;
    }
    public boolean Displayed;

    public Scene(SerialNumber sn, String desc, SerialNumber cap,
                 boolean displayed)
    {
        serialNumber = sn;
        description = desc;
        caption = cap;
        cause = null;
        hascause = false;
        Displayed = displayed;
    }

    public Scene(SerialNumber sn, String desc, SerialNumber cap,
                 SerialNumber cau, boolean displayed)
    {
        serialNumber = sn;
        description = desc;
        caption = cap;
        cause = cau;
        hascause = true;
        Displayed = displayed;
    }

    // Don't instantiate with default constructor!
    Scene(){}
}
